subsys/net/ip/tcp.c

```diff
@@ -2017,6 +2020,42 @@ static void tcp_out_of_order_data(struct tcp *conn, struct net_pkt *pkt,
 	tcp_queue_recv_data(conn, pkt, data_len, seq);
 }
 
+static void tcp_check_sock_options(struct tcp *conn)
+{
+	int sndbuf_opt = 0;
+	int rcvbuf_opt = 0;
+
+	if (IS_ENABLED(CONFIG_NET_CONTEXT_SNDBUF)) {
+		(void)net_context_get_option(conn->context, NET_OPT_SNDBUF,
+					     &sndbuf_opt, NULL);
+	}
+
+	if (IS_ENABLED(CONFIG_NET_CONTEXT_RCVBUF)) {
+		(void)net_context_get_option(conn->context, NET_OPT_RCVBUF,
+					     &rcvbuf_opt, NULL);
+	}
+
+	if (sndbuf_opt > 0 && sndbuf_opt != conn->send_win_max) {
+		k_mutex_lock(&conn->lock, K_FOREVER);
+
+		conn->send_win_max = sndbuf_opt;
+		if (conn->send_win > conn->send_win_max) {
+			conn->send_win = conn->send_win_max;
+		}
+
+		k_mutex_unlock(&conn->lock);
+	}
+
+	if (rcvbuf_opt > 0 && rcvbuf_opt != conn->recv_win_max) {
+		k_mutex_lock(&conn->lock, K_FOREVER);
+
+		conn->recv_win_max = rcvbuf_opt;
```

Comment:
I think this line has to be swapped with the line below, otherwise below (rcvbuf_opt - conn->recv_win_max), will always be 0.

Reply:
Good point, I've made some last minute changes and missed the fact it make no sense now. The current window was limited to `recv_win_max` anyway inside  `tcp_update_recv_wnd()` so the new test did not catch it. I'll restore the previous behaviour (calculate diff before updating `recv_win_max`.